nvflare/apis/stream_shareable.py

```diff
@@ -0,0 +1,163 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC, abstractmethod
+from builtins import dict as StreamMeta
+from typing import Any, Dict, Tuple
+
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable
+
+
+class StreamMetaKey:
+    CHANNEL = "__channel__"
+    TOPIC = "__topic__"
+    RC = "__RC__"
+
+
+class StreamShareableGenerator(ABC):
+    @abstractmethod
+    def get_next(
+        self,
+        stream_meta: StreamMeta,
+        fl_ctx: FLContext,
+    ) -> Tuple[Shareable, float]:
+        """Called to generate next Shareable object to be sent.
+        If this method needs to take long time, it should check the abort_signal in the fl_ctx frequently.
+        If aborted it should return immediately.
+        You can get the abort_signal by calling fl_ctx.get_run_abort_signal().
+
+        Args:
+            stream_meta: stream metadata
+            fl_ctx: The FLContext object
+
+        Returns: a tuple of (Shareable object to be sent, timeout for sending this object)
+
+        """
+        pass
+
+    @abstractmethod
+    def process_replies(
+        self,
+        replies: Dict[str, Shareable],
+        stream_meta: StreamMeta,
+        fl_ctx: FLContext,
+    ) -> Any:
+        """Called to process replies from receivers of the last Shareable object sent to them.
+
+        Args:
+            replies: replies from receivers. It's dict of site_name => reply
+            stream_meta: stream metadata
+            fl_ctx: the FLContext object
+
+        Returns: Any object or None
+
+        If None is returned, the streaming will continue; otherwise the streaming stops and the returned object is
+        returned as the final result of the streaming.
+
+        """
+        pass
+
+
+class StreamShareableProcessor(ABC):
+    @abstractmethod
+    def process(
+        self,
+        shareable: Shareable,
+        stream_meta: StreamMeta,
+        fl_ctx: FLContext,
+    ) -> Tuple[bool, Shareable]:
+        """Process received Shareable object in the stream.
+
+        Args:
+            stream_meta: the stream metadata.
+            shareable: the Shareable object to be processed
+            fl_ctx: the FLContext object
+
+        Returns: a tuple of (whether to continue streaming, reply message)
+
+        Note: the channel and topic here are defined by the app. They are not the regular message headers
+        (CHANNEL and TOPIC) defined in MessageHeaderKey.
+
+        """
+        pass
+
+    def finalize(
+        self,
+        stream_meta: StreamMeta,
+        fl_ctx: FLContext,
+    ):
+        """Called to finalize the generator.
+
+        Args:
+            stream_meta: stream metadata
+            fl_ctx: the FLContext object
+
+        Returns: None
+
+        This method is guaranteed to be called at the end of streaming.
+
+        """
+        pass
+
+
+class StreamShareableProcessorFactory(ABC):
```

Comment:
StreamShareableProcessorPool maybe a better name for this.

Reply:
Will change to ConsumerFactory, which produces an ObjectConsumer object when called.